uClinux-dist/user/ip-sentinel/doc/ip-sentinel.man
.\" $Id: ip-sentinel.man,v 1.1 2005/07/25 03:51:08 magicyang Exp $
.\" Copyright (C) 2002,2003,2004,2005 Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>
.\"  
.\" Permission is granted to make and distribute verbatim copies of
.\" this manual provided the copyright notice and this permission notice
.\" are preserved on all copies.
.\"  
.\" Permission is granted to process this file through groff/nroff and related
.\" tools and print the results, provided the printed document carries copying
.\" permission notice identical to this one except for the removal of this
.\" paragraph (this paragraph not being relevant to the printed manual).
.\"  
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a 
.\" permission notice identical to this one.
.\"  
.\" Permission is granted to copy and distribute translations of this manual
.\" into another language, under the above conditions for modified versions,
.\" except that this permission notice may be stated in a translation
.\" approved by the Free Software Foundation
.\"
..
.de TQ
.br
.ns
.TP \\$1
..
.TH ip-sentinel 8 "Dec 15 2004" "@PACKAGE@ 0.11"
.\"
.\" ====================
.\"
.SH NAME
.PP
ip-sentinel \- keeps your IP space clean
.\"
.\" ====================
.\"
.SH SYNOPSIS
.B ip-sentinel
.RB [ \-\-user | \-u
.IR username ]
.RB [ \-\-group | \-g
.IR groupname ]
.RB [ \-\-chroot | \-r
.IR dir ]
.RB [ \-\-ipfile | \-i
.IR file ]
.RB [ \-\-pidfile | \-p
.IR file ]
.RB [ \-\-logfile | \-l
.IR file ]
.RB [ \-\-errfile | \-e
.IR file ]
.RB [ \-\-mac
.IR mac ]
.RB [ \-\-llmac
.IR mac ]
.RB [ \-\-direction
.IR dir ]
.RB [ \-\-action
.IR program ]
.RB [ \-\-nofork | \-n ]
.RB [ \-\-help | \-h ]
.RB [ \-\-version ]
.IR interface
.\"
.\" ====================
.\"
.SH DESCRIPTION
.B ip-sentinel
starts a daemon which tries to prevent unauthorized usage of IPs
within an ethernet broadcastdomain by answering ARP requests. After
receiving such a faked reply, the requesting party stores the MAC
address in its ARP-table and will send future packets to this invalid
MAC, rendering the IP unreachable.

After startup it opens its logfiles, goes into a chroot jail, drops
privileges and forks itself into background by default. The parameters
of these operations can by changed by the options described below.

.B Warning:
.\"
with an uncareful configuration, ip-sentinel can bring down the entire
network and your physical presence at the ip-sentinel machine and/or
the switch will be required to fix it. Some switches can show an
unexpected behavior are even crash when the special \fI802.*\fR
mac-addresses will be used.
.\"
.\" ====================
.\"
.SH OPTIONS
.TP
.BR \-\-user | \-u\ \fIusername\ \fRand\fB\ \-\-group | \-g\ \fIgroupname
After doing basical initializations, the program will drop its
privilegies by changing its user and group ID. The \fIusername\fR and
\fIgroupname\fR parameters determine these new IDs and must be an
alphanumeric username/groupname or a numeric uid/gid.

.B Attention:
.\"
When having a dietlibc-compiled version in a system which uses remote
NSS (e.g. LDAP, NIS) to make passwd-lookups, you will have to use
numeric IDs.
.\"
.TP
.BR \-\-chroot | \-r\ \fIdirname
Specifies the directory where the program shall go into and call
.B chroot()
after finishing its initialization. By default it is the homedirectory
of the user specified with the \fB\-\-user\fR option.
.TP
.BR \-\-ipfile | \-i\ \fIfilename
Sets the name of the configuration file which contains the list of
forbidden IPs. This name is relative to the chroot and is
.BR ips.cfg
by default. This file will be reread automatically after changes.
.TP
.BR \-\-pidfile | \-p\ \fIfilename
Sets the file where the pid of the daemon will be written into. This
filename is relative to the root-filesystem and is
.BR /var/run/ip\-sentinel.run
by default.
.TP
.BR \-\-logfile | \-l\ \fIfilename
Sets the file where normal operations will be logged. Normal
operations are e.g. an ARP-reply for a forbidden IP. This filename is
relative to the root-filesystem and
.BR /var/log/ip-sentinel.out
by default.
.TP
.BR \-\-errfile | \-e\ \fIfilename
Sets the file where error-conditions will be logged. Error-conditions
are e.g. detected DOS attacks (bombing the ip-sentinel with
ARP-requests) or errors in the configuration file. This filename is
relative to the root-filesystem and
.BR /var/log/ip-sentinel.err
by default.
.TP
.BR \-\-nofork | \-n
Prevents daemon from going into background after initialization. This
option may be useful when using advanced init-concepts like minit
(http://www.fefe.de/minit/).
.TP
.BR \-\-direction\ \fIdir
ip-sentinel can answer both requests asking for an intruder's address,
and requests coming from an intruder asking for official
addresses. This parameter specifies which requests shall be
answered. Possible values are \fIFROM\fR (requests from intruders),
\fITO\fR (requests to intruders) and \fIBOTH\fR. The current default
value is \fITO\fR, but will probably change to \fIBOTH\fR in future
versions.
.TP
.BR \-\-mac\ \fImac
Use \fImac\fR as the default faked mac-address; see below for possible
values. This parameter can be overriden in the ipfile for each
questionable IP. The default value is \fIRANDOM\fR.
.TP
.BR \-\-llmac\ \fImac
Specifies the mac-address which will be used in the link-level headers
when answering requests \fBfrom\fR intruders. Additionally to the
values described below, ip-sentinel understands the \fISAME\fR string,
which means that the address from the ARP-header will be used. When
not using the \fIFROM\fR or \fIBOTH\fR directions, the assigned value
will not have any effect. The default is \fILOCAL\fR.
.TP
.BR \-\-poison
When an intruder sends a request \fBfrom\fR a registered ip address
and this option is set, ip-sentinel will send ARP replies which are
poisoning this ip. When the
.I @!srcmac
syntax is used for the questionable ip, the
.I srcmac
is used as the is-at mac. Else, the mac specified by the
.I --mac
option is used.

This option takes an affect only when
.I --destination BOTH \fRor\fI FROM
is specified.
.TP
.BR \-\-action\ \fIprogram
Execute \fBprogram\fR when a faked ARP reply will be generated. It
will be called with 6 arguments:
.RS
.IP \(bu 2
.IR type
0 when an intruder sent the request, 1 when a protected IP was
requested and -1 else (should never happen)
.IP \(bu 2
.IR spa
the protocol address (IP) of the source
.IP \(bu 2
.IR sha
the hardware address (MAC) of the source
.IP \(bu 2
.IR tpa
the protocol address (IP) of the target
.IP \(bu 2
.IR tha
the hardware address (MAC) of the target
.IP \(bu 2
.IR mac
the generated MAC address
.RE
.RS

The program will be executed only without evaluating its return value
or something else. Please note, that the program will be executed in
the chroot with restricted rights.
.RE
.TP
.BR \-\-help | \-h
Prints help message and exits.
.TP
.BR \-\-version
Prints the version of the program and exits.
.TP
.I interface
Specifies the network-interface where ip-sentinel listens for
ARP-requests and sends ARP-replies. It is a name like "eth0"; aliases
like "eth0:1" are not possible.
.\"
.\" ====================
.\"
.SH CONFIGURATION FILE
The location of the configfile is relatively to the choosen
chroot-directory. Its default is determined at compilation time but it
can be overriden by the \fB\-\-ipfile\fR option.

The content of this file has the following syntax:
.\"
.TP
.IR  ip\  [ mac ]
Specifies an IP which shall be blocked. "Blocked" means that
ARP-replies will be generated telling that the IP address \fIip\fR has
the MAC \fImac\fR. When the MAC parameter is not given, a random one
will be choosen.

.B Attention:
.\"
Assigning a MAC in a switched environment will not have the desired
effect, because the switch will be confused when receiving ARP-replies
for the same IP on different ports (this one of the bad-host and this
one of the ip-sentinel host).

Examples are: "192.168.42.23  11:22:33:44:55:66" or "10.0.0.0".
.\"
.TP
.IB  ip @ [ ! ]srcmac\ [mac]
Similarly to above, but when receiving ARP-requests, the entire
sender-address (IP and MAC) will be compared against the \fIip\fR and
\fIsrcmac\fR values. This setting affects packets \fBfrom\fR intruders
only; ip-sentinel will not generate replies when \fIip\fR is
requested.

When the \fIsrcmac\fR is negated with "\fI!\fR", only addresses from
this ip which are \fBnot\fR having this mac will be matching.

Examples are: "10.0.0.1@0:1:2:3:4:5" or "10.0.0.2@!a:b:c:d:e:f".
.\"
.TP
.BI *@ srcmac\ [mac]
Every host using the srcmac will be blocked unless there is a matching
\fIip@srcmac\fR specification. This option does \fBnot\fR support '!' 
modifiers.
.TP
.IB net / mask\ \fR[ \fImac\fR]
.TQ
.IB net / mask @ [ ! ]srcmac\  \fR[ mac \fR]
Specifies that IPs of an entiry network shall be blocked. See above
for the meaning of "blocked". It is possible to specify a MAC address;
else a random one will be generated.

Examples are: "169.0.0.0/8", "192.168.8.15/255.255.65.31" or "192.168.23.42/26 a:b:c:d:e:f".
.TP
.BI ! ip
.TQ
.BI ! ip @ [ ! ]srcmac
The given IP address \fIip\fR will be ignored. By default, any not
specified address will be ignored but when having blocked netmasks it
may be usefully to allow certain IPs.

An example is: "!192.168.23.42"
.TP
.BI ! net / mask
.TQ
.BI ! net / mask @ [ ! ]srcmac
Tells that IPs of the given network shall be ignored.

Examples are: "!192.168.1.0/255.255.255.0" or "!0.0.0.0/0" (the
default).
.TP
.BI #\  comment\ \fR(and\ \fRempty\ \fRlines)
A comment; will be ignored

.PP
.\"
To be switch-friendly, there are only a few random MACs possible which
are having the format "de:ad:be:ef:00:XX". Within a short timespan
only 32 values are possible for \fIXX\fR.

When having overlapping networks and/or single IPs, this one with the
most specified netmask (count of 1's) takes precedence. When netmasks
are equal, networks which are using the "@srcmac" or "@!srcmac" syntax
are taking precedence over those without source-macs. This "@..." rule
\fBdoes not\fR apply to IPs. The behavior is unspecified when having
overlapping networks with the same count of 1's and "@..." 
specification, or when having duplicate IPs.

.\"
.SS PERFORMANCE
.\"
The lookup of single IPs has a complexity of O(log n) and this of
netmasks a complexity of O(n).

.\"
.SS SPECIAL MAC ADDRESSES
.\"
Beside the usual hex-octets-delimited-by-colons mac addresses,
ip-sentinel understands some special strings both on the commandline
and in the configuration file:
.TP
.BR LOCAL
expands to the mac-address of the used interface
.TP
.BR RANDOM
means a random mac-address which is newly calculated on every usage
.TP
.BR 802.1d
expands to 01:80:C2:00:00:00 which is the "Bridge Group Address".
.TP
.BR 802.3x
expands to 01:80:C2:00:00:01 which is the "IEEE Std. 802.3x Full
Duplex PAUSE operation". This MAC address will be blocked by a lot of
switches and will probably become the default in future versions.
.\"
.PP
The \fI802.*\fR addresses are having a special meaning for some
switches and packets having them as destination-address will be
dropped by the switch instead of flooding all ports. But it depends on
the used switch how/if these macs are honored.

Some switches can show an unexpected behavior are even crash if the
special \fI802.*\fR mac-addresses will be used.
.\"
.SS RANGES
.\"
Except in comments, it is possible to specify ranges everywhere in the
configuration file. These ranges are having the format "{from-to}" or
"{item1,item2,...,itemN}". The first format includes any number
beginning at "from" till "to" (inclusive), while the latter format
expands to the listed items only. The expansion happens on a
line-level and it is possible to use more than one range per line, so
that

.RS 0
192.168.0.{1-3}     0:0:0:0:0:1
.RS 0
192.168.1.{1,3}     0:0:0:0:0:2
.RS 0
192.168.{2,4}.{1-3} 0:0:0:0:0:3

is the same like writing

.RS 0
192.168.0.1         0:0:0:0:0:1
.RS 0
192.168.0.2         0:0:0:0:0:1
.RS 0
192.168.0.3         0:0:0:0:0:1
.RS 0
192.168.1.1         0:0:0:0:0:1
.RS 0
192.168.1.3         0:0:0:0:0:1
.RS 0
192.168.2.1         0:0:0:0:0:3
.RS 0
192.168.2.2         0:0:0:0:0:3
.RS 0
192.168.2.3         0:0:0:0:0:3
.RS 0
192.168.4.1         0:0:0:0:0:3
.RS 0
192.168.4.2         0:0:0:0:0:3
.RS 0
192.168.4.3         0:0:0:0:0:3

Because there can be created very much entries with a single line
(e.g. "{0-255}.{0-255}.{0-255}.{0-255}" would cover the entire IPv4
internet), ranges should be used sparely. When possible, large ranges
should be expressed with netmasks.

.\"
.SS EXAMPLE
.RS 0
0.0.0.0/0                 ## Block anything
.RS 0
!192.168.0.0/24           ## Allow IPs of the form 192.168.0.*
.RS 0
192.168.0.0               ## but block 192.168.0.0
.RS 0
192.168.0.1 a:b:c:d:e:f   ## use a special mac for 192.168.0.1
.RS 0
192.168.0.2 802.1d        ## and 01:80:C2:00:00:00 for 192.168.0.2
.RS 0
10.0.0.1@a:a:a:a:a:a
.RS 0
10.0.0.2@!1:1:1:1:1:1
.RS 0
*@b:b:b:b:b:b             ## block MAC b:b:b:b:b:b regardless of the IP

This setup will not send ARP-replies for the IPs 192.168.0.{3-255} but
when a host tries to use e.g. 169.254.145.213, ip-sentinel will tell
that this IP has a MAC of "de:ad:be:ef:00:XX".

When an intruder is at "10.0.0.1" and uses the mac "a:a:a:a:a:a:", a
faked reply will be generated. Users at the same ip but another mac
will be ignored.

In opposite, users with ip "10.0.0.2" and mac "1:1:1:1:1:1" will be
ignored but intruders with other macs (e.g. "2:2:2:2:2:2") are getting
faked replies. When
.I \-\-poision
is used, ip-sentinel will generate a "10.0.0.2 is at 1:1:1:1:1:1"
arp-reply to a broad address.
.\"
.\" ====================
.\"
.SH SEE ALSO
.BR RFC\ 826 ,
.BR IEEE\ Std\ 802.1D
.\"
.\" ====================
.\"
.SH AUTHOR
Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>


uClinux-dist/release_notes
Release Notes for uClinux for Blackfin Version 2005R4

Version: 2005R4

Date: Dec. 30, 2005
Development Environments

Tool Chain: Toolchain Release 2005R4

u-boot: uboot_1.1.3 Release 2005R4

Host platform: Suse Linux 9.2 or above

Target board: STAMP Board

Note: However, other similar host platforms are also supported, but they are not well tested by now.
Deliverables

Source files

uClinux_dist_2005R4.tar.bz2

Linux ELF file

linux-bf533
linux-bf537
linux-bf561

Compressed Linux image

uImage-bf533
uImage-bf537
uImage-bf561

This document

release_notes_2005R4.pdf

Compressed archive of test results

test_results_bf533_2005R4.tar.gz
test_results_bf537_2005R4.tar.gz
test_results_bf561_2005R4.tar.gz

Summary of test results

test_results_summary_2005R4

Fixes in release 2005R4

See tracker at:

http://blackfin.uclinux.org/tracker/index.php?group_id=17&atid=141
Improvements

   1.
      Support BF561-EZKIT and add a tool to load code to run in core B.
   2.
      Support Bluetechnix CM Platform, which is also based on BF533.
   3.
      Enable module support in default configuration and generate module dependency file for modprobe.
   4.
      Upgrade to mainline source code
          *
            Upgrade uClibc to version 0.9.27.
          *
            Upgrade Linphone to version 1.1.0
          *
            Upgrade LTP to 20051205 release.
   5.
      Reorganize the scheduling and signal handling.
          *
            Defer scheduling and signal handling from different interrupt handlers to interrupt level 15.
          *
            Deal with the deferred exception handler and the defered interrupt handler in the same level of system call handler.
          *
            Run schedule and do signal from each interrupt handler above level 14.
          *
            Move softirq handler to interrupt level 14
   6.
      Prevent double fault exception.
          *
            Make sure that in an exception handler, we use a stack that is covered by a CPLB.
          *
            For exceptions other than CPLB misses, lower our priority to IRQ5, where we can use the normal kernel stack and use other regular kernel functions.
   7.
      Port new applications, libraries and tools
          *
            samba
          *
            strace
          *
            SQLite
          *
            lmbench3.
          *
            whetstone.
          *
            network audio player
          *
            ncurses
   8.
      Work around new hardware anomalys.
          *
            Bypass the first DCLPB fault exception, only handle the second one if it is from the same fault address.
          *
            Do not put SSYNC,CSYNC or IDLE in the second to last instruction of a hardware loop.
          *
            Do not use DMA_RUN bit to detect completion of a DMA transfer.
   9.
      Change paths from bfinnommu to blackfin, per request from Linux Kernel Mailing list.
  10.
      Enable oprofiler support in kernel and Port oprofiler tools
  11.
      Enable interrupt latency benchmark
  12.
      Add new device drivers
          *
            Enable IDE driver
          *
            Enable 802.11 WLAN driver
          *
            Enable PPI driver
          *
            Enable NAND flash driver
          *
            Port YAFFS driver
          *
            Add a common SPI channel driver. All high level drivers access SPI device by this driver.
  13.
      ADI1836 audio driver improvement
          *
            Reduce dummy buffer size from 256 bytes to 4 bytes
          *
            No interrupts generated in dummy DMA transfer
          *
            When audio driver is not active, DMA is also stopped.
          *
            Enable I2S mode for audio input and output. Modify mixer for I2S mode.
          *
            Enable multichannel output. Add a mixer function to change the output channels when play stereo.
          *
            Enable play multi independent audio streams concurrently.
          *
            Make audio DMA stable.
  14.
      Kernel debugging improvement
          *
            Enable hardware single stepping for application debugging.
          *
            Report precise hardware error information
          *
            debug kernel start. Install a mini-execption handler (replacing the U-Boot one), turn the UART on, and dump out the SEQSTAT register.
          *
            Enable hardware trace buffer in dump
  15.
      Improve power management
          *
            Implement the suspend function in bf533_pm_suspend.
          *
            SDRAM enters self-refresh mode before deepsleep to save more power.
          *
            Enable PM on BF537
  16.
      Enable PPPoE
  17.
      Enable DMA in SPI driver.
  18.
      Add board specific Kconfig Option for PF port/pin, which is used for flash and ethernet sharing.
  19.
      Enhance msh to support funtions in scripts.
  20.
      Transmite bfin atomic testandset operation in exception 4 into a new syscall, sys_bfin_spinlock.

Known Issues

A full list of known issues can be found at:

http://blackfin.uclinux.org/tracker/index.php?group_id=17&atid=141
No 	Issue Title
891	tcp stress testing echo service
900	busybox/msh,cd,tar and pipes
960	NFS unstable under heavy load
991	Infrequent audio crash
1012	After running dpmc to_deepsleep testcase, the kernel would panic after a while
1014	kernel built with audio configuration sometimes would fail to boot up and hangs

There are also some issues in the LTP test cases. They are recorded as bug 532, 561, 567, 732, 735, 742, 744, 745, 890 and 1010.
Build Procedure

1. Install Toolchain Release 2005R4

Go to http://blackfin.uclinux.org/frs/?group_id=18 for more information


uClinux-dist/user/ip-sentinel/src/ip-sentinel.c
// $Id: ip-sentinel.c,v 1.1 2005/07/25 03:51:09 magicyang Exp $    --*- c++ -*--

// Copyright (C) 2002,2003,2004,2005 Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>
//  
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; version 2 of the License.
//  
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//  
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
//  

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include "arguments.h"
#include "wrappers.h"
#include "util.h"
#include "parameters.h"
#include "blacklist.h"
#include "antidos.h"
#include "arpmessage.h"
#include "ip-sentinel.h"
#include "worker.h"
#include "jobinfo.h"

#include <assert.h>
#include <unistd.h>
#include <signal.h>
#include <net/if.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <net/if.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/wait.h>
#include <grp.h>

struct ether_addr	local_mac_address;
volatile sig_atomic_t	do_reload;

static void sigHup(int);

void
sigHup(int sig UNUSED)
{
  do_reload = 1;
  signal(SIGHUP, sigHup);
}

static void ALWAYSINLINE
adjustUserGroup(struct Arguments *arguments, uid_t *uid, gid_t *gid)
{
  struct passwd const *	pw_user;
  char *		err_ptr;

  assert(uid!=0 && gid!=0);
  
  *gid = static_cast(gid_t)(-1);
  *uid = strtol(arguments->user, &err_ptr, 0);
  if (arguments->user[0]!='\0' && *err_ptr=='\0') pw_user = getpwuid(*uid);
  else                                            pw_user = Egetpwnam(arguments->user);
  
  if (pw_user!=0) {
    *uid = pw_user->pw_uid;
    *gid = pw_user->pw_gid;
    if (arguments->chroot==0) arguments->chroot = pw_user->pw_dir;
  }

  if (arguments->group!=0) {
    *gid = strtol(arguments->group, &err_ptr, 0);
    if (arguments->group[0]=='\0' || *err_ptr!='\0')
      *gid = Egetgrnam(arguments->group)->gr_gid;
  }

  if (*gid==static_cast(gid_t)(-1)) {
    WRITE_MSGSTR(2, "Failed to determine gid; perhaps you should use the '-g' option. Aborting...\n");
    exit(1);
  }
}

static void //ALWAYSINLINE
daemonize(struct Arguments *arguments)
{
  int			err_fd, out_fd, pid_fd;
  int			aux;
  uid_t			uid;
  gid_t			gid;
  pid_t			daemon_pid;

  err_fd = Eopen(arguments->errfile, O_WRONLY|O_CREAT|O_APPEND|O_NONBLOCK, 0600);
  out_fd = Eopen(arguments->logfile, O_WRONLY|O_CREAT|O_APPEND|O_NONBLOCK, 0600);
  pid_fd = (!arguments->do_fork ? -1 :
	    Eopen(arguments->pidfile, O_WRONLY|O_CREAT|O_TRUNC, 0755));

  adjustUserGroup(arguments, &uid, &gid);
  
  if (arguments->chroot != 0) {
    Echdir(arguments->chroot);
    Echroot(arguments->chroot);
  }

  Esetgroups(1, &gid);
  Esetgid(gid);
  Esetuid(uid);

  Eclose(0);

  aux = Eopen(arguments->ipfile, O_RDONLY, 0);
  Eclose(aux);

  if (arguments->do_fork) daemon_pid = Efork();
  else                    daemon_pid = 0;

  switch (daemon_pid) {
    case 0	:  break;
    default	:
      writeUInt(pid_fd, daemon_pid);
      Ewrite  (pid_fd, "\n", 1);
      assert(daemon_pid!=-1);
      exit(0);
      break;
  }

  if (arguments->do_fork) Esetsid();

  Edup2(out_fd, 1);
  Edup2(err_fd, 2);

  if (pid_fd!=-1) Eclose(pid_fd);
  Eclose(out_fd);
  Eclose(err_fd);

  SETCLOEXEC(1);
  SETCLOEXEC(2);
}

inline static int ALWAYSINLINE
initIfaceInformation(int fd, char const *iface_name)
{
  struct ifreq		iface;
  int			ifidx;
  
  assert(iface_name!=0);
  memcpy(iface.ifr_name, iface_name, IFNAMSIZ);
  Eioctl(fd, SIOCGIFINDEX, &iface);
  if (iface.ifr_ifindex<0) {
    WRITE_MSGSTR(2, "No such interface '");
    WRITE_MSG   (2, iface_name);
    WRITE_MSGSTR(2, "'\n");
    exit(1);
  }

  ifidx = iface.ifr_ifindex; 
  Eioctl(fd, SIOCGIFHWADDR, &iface);
  switch (iface.ifr_hwaddr.sa_family) {
    case ARPHRD_ETHER   :
      memcpy(&local_mac_address, iface.ifr_hwaddr.sa_data, sizeof(local_mac_address));
      break;

    default             :
      WRITE_MSGSTR(2, "unsupported hardware-address (MAC) family of used interface\n");
      exit(1);
  }

  return ifidx;
}

static void ALWAYSINLINE
generateJobToIntruder(struct Worker *worker,
		      AntiDOS *anti_dos, BlackList const *blacklist,
		      ArpMessage const *const msg,
		      unsigned 
int *oversize_sleep)
{
  struct ether_addr const	*mac;
  int				arp_count;
  struct RequestInfo		job;
  struct in_addr const *	src_ip = reinterpret_cast(struct in_addr const *)(msg->data.arp_spa);

  struct BlackListQuery		query = {
    .ip  = reinterpret_cast(struct in_addr const *)(msg->data.arp_tpa),
    .mac = 0,
  };
  
  mac = BlackList_getMac(blacklist, &query);
  if (mac==0) return;
  
  assert(src_ip!=0);
  if (AntiDOS_isOversized(anti_dos)) {
    writeMsgTimestamp(2);
    WRITE_MSGSTR(2, ": Too much requests from too much IPs; last IP was ");
    writeIP     (2, *src_ip);
    WRITE_MSGSTR(2, "\n");
    
    sleep(*oversize_sleep);
    *oversize_sleep = MIN(*oversize_sleep+1, 10);

    return;
  }

  *oversize_sleep = 1;
  arp_count = AntiDOS_registerIP(anti_dos, *src_ip);

  if (isDOS(arp_count)) {
    writeMsgTimestamp(2);
    WRITE_MSGSTR(2, ": Too much requests from ");
    writeIP     (2, *src_ip);
    WRITE_MSGSTR(2, "; DOS-measurement was ");
    writeUInt   (2, arp_count);
    WRITE_MSGSTR(2, "\n");
    return;
  }

  memset(&job, 0, sizeof job);
  job.request      = msg->data;
  job.mac          = *mac;
  job.type         = jobDST;
  job.poison_mac.f = false;

  Worker_sendJob(worker, &job);
}


static void ALWAYSINLINE
generateJobFromIntruder(struct Worker *worker,
			AntiDOS *anti_dos, BlackList const *blacklist,
			ArpMessage const *const msg,
			unsigned int *oversize_sleep)
{
  struct ether_addr const	*mac;
  int				arp_count;
  struct RequestInfo		job;
  struct in_addr const *	dst_ip  = reinterpret_cast(struct in_addr const *)(msg->data.arp_tpa);

  struct BlackListQuery		query = {
    .ip  = reinterpret_cast(struct in_addr const *)(msg->data.arp_spa),
    .mac = reinterpret_cast(struct ether_addr const *)(msg->data.arp_sha),
  };
    
    // Ignore 0.0.0.0 since it is used for duplicate address detection and/or
    // by DHCPDISCOVER
  if (query.ip->s_addr==0) return;
  
  mac = BlackList_getMac(blacklist, &query);
  if (mac==0) return;
  
  assert(dst_ip!=0);
  if (AntiDOS_isOversized(anti_dos)) {
    writeMsgTimestamp(2);
    WRITE_MSGSTR(2, ": Too much requests from too much IPs; last IP was ");
    writeIP     (2, *dst_ip);
    WRITE_MSGSTR(2, "\n");
    
    sleep(*oversize_sleep);
    *oversize_sleep = MIN(*oversize_sleep+1, 10);
    return;
  }

  *oversize_sleep = 1;
  arp_count = AntiDOS_registerIP(anti_dos, *query.ip);

  if (isDOS(arp_count)) {
    writeMsgTimestamp(2);
    WRITE_MSGSTR(2, ": Too much requests from intruder ");
    writeIP     (2, *query.ip);
    WRITE_MSGSTR(2, "; DOS-measurement was ");
    writeUInt   (2, arp_count);
    WRITE_MSGSTR(2, ", current dst ");
    writeIP     (2, *dst_ip);
    WRITE_MSGSTR(2, "\n");
    return;
  }

  memset(&job, 0, sizeof job);
  job.request = msg->data;
  job.mac     = *mac;
  job.type    = jobSRC;

  if (query.poison_mac) {
    job.poison_mac.f = true;
    job.poison_mac.v = *query.poison_mac;
  }
  else
    job.poison_mac.f = false;

  Worker_sendJob(worker, &job);
}


static void NORETURN //ALWAYSINLINE
run(struct Worker *worker, struct Arguments const *args) 
{
  BlackList			cfg;
  AntiDOS			anti_dos;
  unsigned int			error_count = 0;
  struct sockaddr_ll		addr;
  socklen_t			from_len;
  char				buffer[4096];
  ArpMessage const * const	msg    = reinterpret_cast(ArpMessage const *)(buffer);
  unsigned int			oversize_sleep = 1;
  int				sock = worker->sock;

  memset(&addr, 0, sizeof(addr));
  
  BlackList_init(&cfg, args);
  AntiDOS_init(&anti_dos);

  while (true) {
    size_t			size;
    
    AntiDOS_update(&anti_dos);

    from_len = sizeof(addr);
    size     = TEMP_FAILURE_RETRY(recvfrom(sock, buffer, sizeof buffer, 0,
					   (struct sockaddr *)(&addr), &from_len));

    if (static_cast(ssize_t)(size)==-1) {
      ++error_count;
      if (error_count>MAX_ERRORS) {
	perror("recvfrom()");
	exit(1);
      }

      continue;
    }

    error_count = 0;

    if (ntohs(addr.sll_protocol)     !=ETHERTYPE_ARP) continue;
    if (ntohs(msg->data.ea_hdr.ar_op)!=ARPOP_REQUEST) continue;

    if (!do_reload) BlackList_softUpdate(&cfg);
    else {
      BlackList_update(&cfg);
      do_reload = false;
    }

    if (args->arp_dir & dirTO)
      generateJobToIntruder  (worker, &anti_dos, &cfg, msg, &oversize_sleep);

    if (args->arp_dir & dirFROM)
      generateJobFromIntruder(worker, &anti_dos, &cfg, msg, &oversize_sleep);
  }
}

inline static int ALWAYSINLINE
generateSocket(char const *iface, int *idx)
{
  int			sock = Esocket(PF_PACKET, SOCK_RAW, htons(ETH_P_ARP));
  struct sockaddr_ll	addr;

  assert(iface!=0);
  assert(idx!=0);

  SETCLOEXEC(sock);
  *idx = initIfaceInformation(sock, iface);

  memset(&addr, 0, sizeof(addr));
  addr.sll_family   = AF_PACKET;
  addr.sll_protocol = htons(ETH_P_ARP);
  addr.sll_ifindex  = *idx;
  
  (void)Ebind(sock, &addr, sizeof addr);

  return sock;
}

int
main(int argc, char *argv[])
{
  struct Arguments	arguments;
  int			sock;
  int			if_idx;
  struct Worker		worker;
  
  parseOptions(argc, argv, &arguments);

  sock   = generateSocket(arguments.iface, &if_idx);
  daemonize(&arguments);

  Arguments_fixupOptions(&arguments);
  
  Worker_init(&worker, &arguments, sock, if_idx);

  signal(SIGHUP,  sigHup);

  run(&worker, &arguments);
}
